motionplan/constraint.go

```diff
@@ -10,120 +10,228 @@ import (
+// Return values are:
+// -- a bool representing whether all constraints passed
+// -- if failing, a string naming the failed constraint.
+func (c *ConstraintHandler) CheckSegmentConstraints(cInput *SegmentInput) (bool, string) {
+	for name, cFunc := range c.segmentConstraints {
+		pass := cFunc(cInput)
+		if !pass {
+			return false, name
+		}
+	}
+	return true, ""
+}
+
+// CheckStateConstraintsAcrossSegment will interpolate the given input from the StartInput to the EndInput, and ensure that all intermediate
+// states as well as both endpoints satisfy all state constraints. If all constraints are satisfied, then this will return `true, nil`.
+// If any constraints fail, this will return false, and an SegmentInput representing the valid portion of the segment, if any. If no
+// part of the segment is valid, then `false, nil` is returned.
+func (c *ConstraintHandler) CheckStateConstraintsAcrossSegment(ci *SegmentInput, resolution float64) (bool, *SegmentInput) {
+	// ensure we have cartesian positions
+	err := ci.resolveInputsToPositions()
 	if err != nil {
 		return false, nil
 	}
-	steps := PathStepCount(ci.StartPos, ci.EndPos, resolution)
+	steps := PathStepCount(ci.StartPosition, ci.EndPosition, resolution)
 
 	var lastGood []referenceframe.Input
-	// Seed with just the start position to walk the path
-	interpC := &ConstraintInput{Frame: ci.Frame}
-	interpC.StartInput = ci.StartInput
-	interpC.EndInput = ci.StartInput
 
-	for i := 1; i <= steps; i++ {
+	for i := 0; i <= steps; i++ {
 		interp := float64(i) / float64(steps)
-		interpC, err = cachedInterpolateInput(ci, interp, interpC.EndInput, interpC.EndPos)
+		interpConfig := referenceframe.InterpolateInputs(ci.StartConfiguration, ci.EndConfiguration, interp)
+		interpC := &StateInput{Frame: ci.Frame, Configuration: interpConfig}
+		err = interpC.resolveInputsToPositions()
 		if err != nil {
 			return false, nil
 		}
-		pass, _, _ := c.CheckConstraints(interpC)
+		pass, _ := c.CheckStateConstraints(interpC)
 		if !pass {
-			if i > 1 {
-				return false, &ConstraintInput{StartInput: lastGood, EndInput: interpC.StartInput}
+			if i == 0 {
+				// fail on start pos
+				return false, nil
 			}
-			// fail on start pos
-			return false, nil
+			return false, &SegmentInput{StartConfiguration: ci.StartConfiguration, EndConfiguration: lastGood}
 		}
-		lastGood = interpC.StartInput
+		lastGood = interpC.Configuration
 	}
-	// extra step to check the end
-	if err != nil {
+
+	return true, nil
+}
+
+// CheckSegmentAndStateValidity will check an segment input and confirm that it 1) meets all segment constraints, and 2) meets all
+// state constraints across the segment at some resolution. If it fails an intermediate state, it will return the shortest valid segment,
+// provided that segment also meets segment constraints.
+func (c *ConstraintHandler) CheckSegmentAndStateValidity(cInput *SegmentInput, resolution float64) (bool, *SegmentInput) {
```

Comment:
Without a concrete idea of how a `StateConstraint` and a `SegmentConstraint` might differ, this question may be silly, but bear with me: Would it be helpful to differentiate between what failed across the calls to `CheckStateConstraintsAcrossSegment` and `CheckSegmentConstraints` before returning in this function?

Reply:
I'm not sure I fully understand the question.

But your comment made me look at this function again and realize the initial segment check is useless, or should at least be put off until after the segment check.

Reply:
Basically, is there any reason to highlight that the overall `SegmentConstraint` is what caused a problem versus a specific set of `StateConstraints` within a segment?

Reply:
Sounds good, thanks!